docs: Getting started rewrite (#16497)
grafana/loki: docs/sources/get-started/quick-start.md

Provide feedback, report bugs, and raise issues in the [Grafana Killercoda repository](https://github.com/grafana/killercoda).
{{< /admonition >}}
<!-- INTERACTIVE ignore END -->

<!-- INTERACTIVE page intro.md END -->

<!-- INTERACTIVE page step1.md START -->

## Deploy the Loki Stack

<!-- INTERACTIVE ignore START -->
{{< admonition type="note" >}}
This quickstart assumes you are running Linux or MacOS. Windows users can follow the same steps using [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/install).
{{< /admonition >}}
<!-- INTERACTIVE ignore END -->

**To deploy the Loki Stack locally, follow these steps:**

1. Clone the Loki fundamentals repository and checkout the getting-started branch:

     ```bash
     git clone https://github.com/grafana/loki-fundamentals.git -b getting-started
     ```

1. Change to the `loki-fundamentals` directory:

     ```bash
     cd loki-fundamentals
     ```

1. With `loki-fundamentals` as the current working directory deploy Loki, Alloy, and Grafana using Docker Compose:

     ```bash
     docker compose up -d
     ```
      After running the command, you should see a similar output:

      ```console

>        ✔ Container loki-fundamentals-grafana-1  Started  0.3s 

Reviewer: I'm not seeing the Drilldown menu item in Killercoda. Drilldown is still under Explore > Logs.  And the direct link at line 103 gives a 404.

And when I run it on my local, I'm getting the same result.  We've somehow lost the Drilldown menu that we had the last time I ran through this.

Loki by design does not force log lines into a specific schema format. Whether you are using JSON, key-value pairs, plain text, Logfmt, or any other format, Loki ingests these logs lines as a stream of characters. The sample application we are using stores logs in [Logfmt](https://brandur.org/logfmt) format:
<!-- INTERACTIVE copy START -->
```bash
ts=2025-02-21 16:09:42,176 level=INFO line=97 msg="192.168.65.1 - - [21/Feb/2025 16:09:42] "GET /static/style.css HTTP/1.1" 304 -"
```
To break this down:
- `ts=2025-02-21 16:09:42,176` is the timestamp of the log line.
- `level=INFO` is the log level.
- `line=97` is the line number in the code.
- `msg="192.168.65.1 - - [21/Feb/2025 16:09:42] "GET /static/style.css HTTP/1.1" 304 -"` is the log message.

<!-- INTERACTIVE copy END -->
When querying Loki, you can pipe the result of the label selector through a parser. This extracts attributes from the log line for further processing. For example, lets pipe `{container="greenhouse-main_app-1"}` through the `logfmt` parser to extract the `level` and `line` attributes:
<!-- INTERACTIVE copy START -->
```logql
{container="greenhouse-main_app-1"} | logfmt
```
<!-- INTERACTIVE copy END -->

> When you now expand a log line in the query result, you will see the extracted attributes.

Reviewer: Since this is a getting started, let's not assume that users know about log details yet.
```suggestion
When you click the arrow (>) next to a log line in the query result to expand the view, you will see the extracted attributes in the log details.
```

> {{< admonition type="tip" >}}

Reviewer: I don't know if we want to move this tip higher up the page to give the service more time to generate errors, or add a line about waiting a minute or two, but a couple of times when I've run the first query, I haven't had any data that matched the query.  When I run it again, there's usually a result, but sometimes it takes a minute or two to generate enough errors to get data from my query.

<!-- INTERACTIVE copy START -->
```logql
{container="greenhouse-plant_service-1"} | logfmt | level="ERROR"
```
<!-- INTERACTIVE copy END -->
This query will return all the logs from the `greenhouse-plant_service-1` container that have a `level` attribute of `ERROR`. You can further refine this query by filtering for a specific code line:
<!-- INTERACTIVE copy START -->
```logql
{container="greenhouse-plant_service-1"} | logfmt | level="ERROR", line="58"
```
<!-- INTERACTIVE copy END -->
This query will return all the logs from the `greenhouse-plant_service-1` container that have a `level` attribute of `ERROR` and a `line` attribute of `58`.

LogQL also supports metrics queries. Metrics are useful for abstracting the raw log data aggregating attributes into numeric values. This allows you to utilise more visualization options in Grafana as well as generate alerts on your logs. 

For example, you can use a metric query to count the number of logs per second that have a specific attribute:
<!-- INTERACTIVE copy START -->
```logql

> sum(rate({container="greenhouse-plant_service-1"} | logfmt | level="ERROR" [$__auto]))

Reviewer: The data from this query is hard to see in the graph when it's set to `lines` since it's just a couple of green dots.  It might be helpful to add another sentence here and direct users to change the graph view to `bars` to make the data more visible in the graph.  That would also teach new users that there are multiple graph visualizations available that they can play around with.

```logql
topk(10,sum(rate({level="error"} | logfmt [5m])) by (service_name))
```
<!-- INTERACTIVE copy END -->
{{< admonition type="note" >}}
`service_name` is a label created by Loki when no service name is provided in the log line. It will use the container name as the service name. A list of all automatically generated labels can be found in [Labels](https://grafana.com/docs/loki/latest/get-started/labels/#default-labels-for-all-users).
{{< /admonition >}}

Finally, lets take a look at the total log throughput of each container in our production environment:
<!-- INTERACTIVE copy START -->
```logql
sum by (service_name) (rate({env="production"} | logfmt [$__auto]))
```
<!-- INTERACTIVE copy END -->
This is made possible by the `service_name` label and the `env` label that we have added to our log lines. Note that `env` is a static label that we added to all log lines as they are processed by Alloy.

<!-- INTERACTIVE page step5.md END -->

<!-- INTERACTIVE page step6.md START -->

## A look under the hood

At this point you will have a running Loki Stack and a sample application generating logs. You have also queried Loki using Grafana Logs Drilldown and Grafana Explore. 

> In this next section we will take a look under the hood to understand how the Loki Stack has been configured to collect logs, the Loki configuration file, and how the Loki datasource has been configured in Grafana.

Reviewer: ```suggestion
In this next section we will take a look under the hood to understand how the Loki Stack has been configured to collect logs, the Loki configuration file, and how the Loki data source has been configured in Grafana.
```

### Grafana Alloy configuration

> Grafana Alloy is collecting logs from all the docker containers and forwarding them to Loki. 

Reviewer: ```suggestion
Grafana Alloy is collecting logs from all the Docker containers and forwarding them to Loki. 
```

It needs a configuration file to know which logs to collect and where to forward them to. Within the `loki-fundamentals` directory, you will find a file called `config.alloy`:

```alloy

> // This component is responsible for disovering new containers within the docker environment

Reviewer: ```suggestion
// This component is responsible for discovering new containers within the Docker environment
```

loki.source.docker "getting_started" {
	host             = "unix:///var/run/docker.sock"
	targets          = discovery.docker.getting_started.targets
	forward_to       = [loki.process.getting_started.receiver]
	relabel_rules    = discovery.relabel.getting_started.rules
	refresh_interval = "5s"
}

// This component is responsible for processing the logs (In this case adding static labels)
loki.process "getting_started" {
    stage.static_labels {
    values = {
      env = "production",
    }
}
    forward_to = [loki.write.getting_started.receiver]
}

// This component is responsible for writing the logs to Loki
loki.write "getting_started" {
	endpoint {
		url  = "http://loki:3100/loki/api/v1/push"
	}
}

// Enables the ability to view logs in the Alloy UI in realtime
livedebugging {
  enabled = true
}
```
This configuration file can be viewed visually via the Alloy UI at [http://localhost:12345/graph](http://localhost:12345/graph).

{{< figure max-width="100%" src="/media/docs/loki/getting-started-alloy-ui.png" caption="Alloy UI" alt="Alloy UI" >}}

In this view you can see the components of the Alloy configuration file and how they are connected:

> * **discovery.docker**: This component queries the metadata of the docker environment via the docker socket and discovers new containers, as well as providing metadata about the containers.

Reviewer: ```suggestion
* **discovery.docker**: This component queries the metadata of the Docker environment via the Docker socket and discovers new containers, as well as providing metadata about the containers.
```

```yaml
auth_enabled: false

server:
  http_listen_port: 3100
  grpc_listen_port: 9096
  log_level: info
  grpc_server_max_concurrent_streams: 1000

common:
  instance_addr: 127.0.0.1
  path_prefix: /tmp/loki
  storage:
    filesystem:
      chunks_directory: /tmp/loki/chunks
      rules_directory: /tmp/loki/rules
  replication_factor: 1
  ring:
    kvstore:
      store: inmemory

query_range:
  results_cache:
    cache:
      embedded_cache:
        enabled: true
        max_size_mb: 100

limits_config:
  metric_aggregation_enabled: true
  allow_structured_metadata: true
  volume_enabled: true
  retention_period: 24h   # 24h

schema_config:
  configs:
    - from: 2020-10-24
      store: tsdb
      object_store: filesystem
      schema: v13
      index:
        prefix: index_
        period: 24h

pattern_ingester:
  enabled: true
  metric_aggregation:
    loki_address: localhost:3100

ruler:
  enable_alertmanager_discovery: true
  enable_api: true
  
frontend:
  encoding: protobuf

compactor:
  working_directory: /tmp/loki/retention
  delete_request_store: filesystem
  retention_enabled: true
```
To summarize the configuration file:

> * **auth_enabled**: This is set to false, meaning Loki does not need a [tenant ID](https://grafana.com/docs/loki/<LOKI_VERSION>/operations/multi-tenancy/) for ingest or query. Note that this is not recommended for production environments. When deploying the Loki helm chart, this is set to true by default.

Reviewer: ```suggestion
* **auth_enabled**: This is set to false, meaning Loki does not need a [tenant ID](https://grafana.com/docs/loki/<LOKI_VERSION>/operations/multi-tenancy/) for ingest or query. Note that this is not recommended for production environments. When deploying the Loki Helm chart, this is set to true by default.
```

* **schema_config**: Defines the schema configuration for Loki. This includes the schema version, the object store, and the index configuration.
* **pattern_ingester**: Enables pattern ingesters which are used to discover log patterns. Mostly used by Grafana Logs Drilldown.
* **ruler**: Enables the ruler component of Loki. This is used to create alerts based on log queries.
* **frontend**: Defines the encoding format for the frontend. In this case it is set to `protobuf`.
* **compactor**: Defines the compactor configuration. Used to compact the index and mange chunk retention.

The above configuration file is a basic configuration file for Loki. For more advanced configuration options, refer to the [Loki Configuration](https://grafana.com/docs/loki/<LOKI_VERSION>/configuration/) documentation.

<!-- INTERACTIVE page step7.md END -->

<!-- INTERACTIVE page step8.md START -->

> ### Grafana Loki Datasource

Reviewer: Global, data source is standardized as two words.
https://grafana.com/docs/writers-toolkit/write/style-guide/word-list/#data-source
```suggestion
### Grafana Loki Data source
```

The final piece of the puzzle is the Grafana Loki datasource. This is used by Grafana to connect to Loki and query the logs. Grafana has multiple ways to define a datasource;
* **Direct**: This is where you define the datasource in the Grafana UI.
* **Provisioning**: This is where you define the datasource in a configuration file and have Grafana automatically create the datasource.
* **API**: This is where you use the Grafana API to create the datasource.

> In this case we are using the provisioning method. Instead of mounting the Grafana configuration directory, we have defined the datasource in the `docker-compose.yml` file:

Reviewer: ```suggestion
In this case we are using the provisioning method. Instead of mounting the Grafana configuration directory, we have defined the data source in this portion of the `docker-compose.yml` file:
```

         apiVersion: 1
         datasources:
         - name: Loki
           type: loki
           access: proxy
           orgId: 1
           url: 'http://loki:3100'
           basicAuth: false
           isDefault: true
           version: 1
           editable: true 
         EOF
         /run.sh
    networks:
      - loki
```
Within the entrypoint section of the `docker-compose.yml` file, we have defined a file called `run.sh` this runs on startup and creates the datasource configuration file `ds.yaml` in the Grafana provisioning directory. 
This file defines the Loki datasource and tells Grafana to use it. Since Loki is running in the same Docker network as Grafana, we can use the service name `loki` as the URL.

<!-- INTERACTIVE page step8.md END -->

<!-- INTERACTIVE page finish.md START -->

## What next?

{{< docs/ignore >}}
### Back to docs
Head back to where you started from to continue with the Loki documentation: [Loki documentation](https://grafana.com/docs/loki/latest/get-started/quick-start/).
{{< /docs/ignore >}}

You have completed the Loki Quickstart demo. So where to go next? Here are a few suggestions:

> * **Deploy:** Loki can be deployed in multiple ways. For production usecases we recommend deploying Loki via the [Helm chart](https://grafana.com/docs/loki/<LOKI_VERSION>/setup/install/helm/).

Reviewer: ```suggestion
* **Deploy:** Loki can be deployed in multiple ways. For production use cases we recommend deploying Loki via the [Helm chart](https://grafana.com/docs/loki/<LOKI_VERSION>/setup/install/helm/).
```

   If you installed in a different directory, you’ll need to modify these queries to match your installation directory.
   Here are some sample queries to get you started using LogQL. After copying any of these queries into the query editor, click **Run Query** (4) to execute the query.

   After copying any of these queries into the query editor, click **Run Query** (4) to execute the query.

   1. View all the log lines which have the container label `evaluate-loki-flog-1`:
   1. View all the log lines which have the `container` label value `greenhouse-main_app-1`:
      <!-- INTERACTIVE copy START -->
      ```bash
      {container="evaluate-loki-flog-1"}
      {container="greenhouse-main_app-1"}
      ```
      <!-- INTERACTIVE copy END -->
      In Loki, this is a log stream.

      Loki uses [labels](https://grafana.com/docs/loki/<LOKI_VERSION>/get-started/labels/) as metadata to describe log streams.

      Loki queries always start with a label selector.
      In the previous query, the label selector is `{container="evaluate-loki-flog-1"}`.
      In the previous query, the label selector is `{container="greenhouse-main_app-1"}`.

>    1. To view all the log lines which have the container label `evaluate-loki-grafana-1`:

Reviewer: I'm sad that we've lost these other examples in the revision.  I was trying to build in some simple LogQL lessons with them.

Author: hmm, let me work on that. I wonder if we can still achieve the same with the containers at our disposal.

> ### Extracting Attributes from Logs

Reviewer: ```suggestion
### Extracting attributes from logs
```

<!-- INTERACTIVE copy START -->
```bash
ts=2025-02-21 16:09:42,176 level=INFO line=97 msg="192.168.65.1 - - [21/Feb/2025 16:09:42] "GET /static/style.css HTTP/1.1" 304 -"
```
To break this down:
- `ts=2025-02-21 16:09:42,176` is the timestamp of the log line.
- `level=INFO` is the log level.
- `line=97` is the line number in the code.
- `msg="192.168.65.1 - - [21/Feb/2025 16:09:42] "GET /static/style.css HTTP/1.1" 304 -"` is the log message.

<!-- INTERACTIVE copy END -->
When querying Loki, you can pipe the result of the label selector through a formatter. This extracts attributes from the log line for further processing. For example lets pipe `{container="greenhouse-main_app-1"}` through the `logfmt` formatter to extract the `level` and `line` attributes:
<!-- INTERACTIVE copy START -->
```bash
{container="greenhouse-main_app-1"} | logfmt
```
<!-- INTERACTIVE copy END -->
When you now expand a log line in the query result, you will see the extracted attributes.

{{< admonition type="tip" >}}

> **Before we move on** to the next section, let's generate some error logs. To do this, enable the bug service in the sample application. This is done by setting the `Toggle Error Mode` to `On` in the Carnivorous Greenhouse application. This will cause the bug service to randomly cause services to fail.

Reviewer: I don't think we need the tip or bold formatting here.

LogQL also supports metrics queries. Metrics are useful for abstracting the raw log data aggregating attributes into numeric values. This allows you to utilise more visualization options in Grafana as well as generate alerts on your logs. 

For example, you can use a metric query to count the number of logs per second that have a specific attribute:
<!-- INTERACTIVE copy START -->
```bash
sum(rate({container="greenhouse-plant_service-1"} | logfmt | level="ERROR" [$__auto]))
```
<!-- INTERACTIVE copy END -->

Another example is to get the top 10 services producing the highest rate of errors:
<!-- INTERACTIVE copy START -->
```bash
topk(10,sum(rate({level="error"} | logfmt [5m])) by (service_name))
```
<!-- INTERACTIVE copy END -->
{{< admonition type="note" >}}
`service_name` is a label created by Loki when no service name is provided in the log line. It will use the container name as the service name. A list of all labels can be found in [Labels](https://grafana.com/docs/loki/latest/get-started/labels/#default-labels-for-all-users).
{{< /admonition >}}

> Lastly, lets take a look at the total log throughput of each container in our production environment:

Reviewer: ```suggestion
Finally, lets take a look at the total log throughput of each container in our production environment:
```
